core/src/main/scala/org/apache/spark/api/python/PythonWorkerFactory.scala

```diff
@@ -60,6 +66,12 @@ private[spark] class PythonWorkerFactory(pythonExec: String, envVars: Map[String
     envVars.getOrElse("PYTHONPATH", ""),
     sys.env.getOrElse("PYTHONPATH", ""))
 
+
+  if (virtualEnvEnabled) {
+    val virtualEnvFactory = new VirtualEnvFactory(pythonExec, conf, false)
+    pythonExec = virtualEnvFactory.setupVirtualEnv()
```

Comment:
Rather than updating a var what about if we created a new val which is set to the value passed in if virtualEnvEnabled is false or this new value if its true?

Reply:
Fixed